src/boot/shim.c

```diff
@@ -116,3 +116,21 @@ void shim_retain_protocol(void) {
          * Requires Shim 15.8. */
         (void) efivar_set_raw(MAKE_GUID_PTR(SHIM_LOCK), u"ShimRetainProtocol", &value, sizeof(value), 0);
 }
+
+typedef struct SHIM_IMAGE_LOADER {
+        void* LoadImage;
+        void* StartImage;
+        void* Exit;
+        void* UnloadImage;
+} SHIM_IMAGE_LOADER;
+
+#define SHIM_IMAGE_LOADER_GUID \
+        { 0x1f492041, 0xfadb, 0x4e59, {0x9e, 0x57, 0x7c, 0xaf, 0xe7, 0x3a, 0x55, 0xab } }
+
+bool shim_load_image_is_shim(void) {
+        EFI_STATUS err;
+        SHIM_IMAGE_LOADER *shim_image_loader;
+
+        err = BS->LocateProtocol(MAKE_GUID_PTR(SHIM_IMAGE_LOADER), NULL, (void **) &shim_image_loader);
+        return err == EFI_SUCCESS && shim_image_loader->LoadImage == BS->LoadImage;
```

Comment:
If the protocol is available, then BS->loadimage will be overridden, so we don't need to explicitly check. Maybe we can get https://github.com/systemd/systemd/pull/37387 merged first, and then you can use the same helper I am adding there

Reply:
I would prefer to check that we are not after a boot loader that decided to provide a different LoadImage that is not working like we expect. At least we should fail.

Reply:
And there is a way for a boot loader to wrap load image and we still obey it. They have to set it in the SHIM_IMAGE_LOAD protocol. This is the contract that they know their implementation works correctly. If not, we have no idea.

Reply:
If you rebase, there's a new helper merged in main now to use or extend.

Given we are not aware of any bootloader actually doing that, I'd prefer if we kept the changes smaller until such case actually appears, at which point we can certainly add it. But I have a feeling others would be much happier if anything that is related to shim is as minimized as possible, given in practice this additional code right now is not necessary.

Reply:
I will rebase in the beginning of the week and try to use it.